aws-cpp-sdk-core/source/internal/AWSHttpResourceClient.cpp

```diff
@@ -72,40 +101,67 @@ Aws::String AWSHttpResourceClient::GetResource(const char* endpoint, const char*
     ss << endpoint << resource;
     AWS_LOGSTREAM_TRACE(m_logtag.c_str(), "Retrieving credentials from " << ss.str().c_str());
 
-    std::shared_ptr<HttpRequest> request(CreateHttpRequest(ss.str(), HttpMethod::HTTP_GET,
-                Aws::Utils::Stream::DefaultResponseStreamFactoryMethod));
-
-    if(authToken)
+    for (long retries = 0;; retries++)
     {
-        request->SetHeaderValue(Aws::Http::AWS_AUTHORIZATION_HEADER, authToken);
-    }
+        std::shared_ptr<HttpRequest> request(CreateHttpRequest(ss.str(), HttpMethod::HTTP_GET,
+                    Aws::Utils::Stream::DefaultResponseStreamFactoryMethod));
 
-    std::shared_ptr<HttpResponse> response(m_httpClient->MakeRequest(request));
+        if (authToken)
+        {
+            request->SetHeaderValue(Aws::Http::AWS_AUTHORIZATION_HEADER, authToken);
+        }
 
-    if (response == nullptr)
-    {
-        AWS_LOGSTREAM_ERROR(m_logtag.c_str(), "Http request to retrieve credentials failed.");
-    }
-    else if (response->GetResponseCode() != HttpResponseCode::OK)
-    {
-        AWS_LOGSTREAM_ERROR(m_logtag.c_str(), "Http request to retrieve credentials failed with error code "
-                << static_cast<int>(response->GetResponseCode()));
-    }
-    else
-    {
-        Aws::IStreamBufIterator eos;
-        return Aws::String(Aws::IStreamBufIterator(response->GetResponseBody()), eos);
-    }
+        std::shared_ptr<HttpResponse> response(m_httpClient->MakeRequest(request));
+
+        const HttpResponseCode responseCode = response ? response->GetResponseCode() : HttpResponseCode::REQUEST_NOT_MADE;
 
-    return "";
+        if (responseCode == HttpResponseCode::OK)
+        {
+            Aws::IStreamBufIterator eos;
+            return Aws::String(Aws::IStreamBufIterator(response->GetResponseBody()), eos);
+        }
+        else if (responseCode == HttpResponseCode::REQUEST_NOT_MADE)
+        {
+            AWS_LOGSTREAM_ERROR(m_logtag.c_str(), "Http request to retrieve credentials failed.");
+        }
+        else if (response->GetResponseCode() >= HttpResponseCode::INTERNAL_SERVER_ERROR)
+        {
+            AWS_LOGSTREAM_ERROR(m_logtag.c_str(), "Http request to retrieve credentials failed with error code "
+                                << static_cast<int>(response->GetResponseCode()));
+        }
+        else
+        {
+            // Client error, retrying the same request will not help
+            AWS_LOGSTREAM_ERROR(m_logtag.c_str(), "Http request to retrieve credentials failed with non-retriable error code "
+                                << static_cast<int>(response->GetResponseCode()));
+            return {};
+        }
+
+        // Internal error that is retryable
+        Aws::Client::AWSError<Aws::Client::CoreErrors> error(Aws::Client::CoreErrors::INTERNAL_FAILURE, true);
```

Comment:
I think it's better to refer `isRetryableHttpResponseCode` in AWSClient.cpp. That's a static function though. But we can update that.

Reply:
I don't get it. I thought that the second parameter to `AWSError` specifies whether it is a retryable error or not.

Reply:
TBH, I don't get the whole design of the error/retries handling (for me it seems more complicated that it should be). It's unclear who is responsible for making decisions about retrying. From one perspective, the `AWSError` has `isRetryable` field that tells whether the error is retryable or not and it seems pretty valid because it is set by a particular service who perfectly knows what errors are retryable. From other perspective there is `RetryStrategy::ShouldRetry` which is responsible for final decision about retrying but none of the standard strategies has any logic beyond checking `AWSError::isRetryable` and this part seems not really needed.

Reply:
Seems I'm getting it. Do you actually mean to replace condition `if (response->GetResponseCode() >= HttpResponseCode::INTERNAL_SERVER_ERROR)` with `if (isRetryableHttpResponseCode(responseCode))`?

Reply:
Btw, there is also `CoreErrorsMapper::GetErrorForHttpResponseCode` functions. One seems to have more correct logic than `isRetryableHttpResponseCode` and it is not static. Should not I use it instead?

Reply:
You bring up a good point. These two functions are actually not consistent with each other. We need to resolve this.